bliss/generate.py

```diff
@@ -0,0 +1,98 @@
+    )  # ceil division
+
+    # stores details of the written image files - { filename: string, data }
+    data_files: List[Dict] = []
+
+    # assume overwriting existing files (if exist)
+    # use SimulatedDataset to generate data in minibatches iteratively,
+    # then concatenate before caching to disk via pickle
+    simulated_dataset = simulator.train_dataloader().dataset
+    assert isinstance(
+        simulated_dataset, IterableDataset
+    ), "simulated_dataset must be IterableDataset"
+
+    # create cached_data_path if it doesn't exist
+    if not os.path.exists(cached_data_path):
+        os.makedirs(cached_data_path)
+
+    if len(data_files) != n_files:
+        # if no written image files, render (and write to disk) images
+        for file_idx in tqdm(range(n_files), desc="Generating and writing cached dataset"):
+            batch_data = []
+            for _ in range(file_data_size // simulator.image_prior.batch_size):
+                batch_data.append(next(iter(simulated_dataset)))
+
+            # TODO: refactor/optimize this dictionary/tensor flattening
+            FileDatum = TypedDict(  # noqa: N806
+                "FileDatum",
+                {"tile_catalog": TileCatalog, "images": torch.Tensor, "background": torch.Tensor},
+            )
+            file_data: List[FileDatum] = []
+            flat_data = {}
+            # concatenate tensors in tile_catalog dictionaries
+            tile_catalog_flattened = {}
+            for key in batch_data[0]["tile_catalog"].keys():
+                tile_catalog_flattened[key] = (
+                    torch.flatten(
+                        torch.cat([data["tile_catalog"][key] for data in batch_data]),
+                        start_dim=0,
+                        end_dim=1,
+                    )
+                    if len(batch_data) > 1
+                    else batch_data[0]["tile_catalog"][key]
+                )
+                flat_data["tile_catalog"] = tile_catalog_flattened
+            for key in batch_data[0].keys():
+                if key != "tile_catalog":
+                    flat_data[key] = (
+                        torch.flatten(
+                            torch.cat([data[key] for data in batch_data]), start_dim=0, end_dim=1
+                        )
+                        if len(batch_data) > 1
+                        else batch_data[0][key]
+                    )
+            assert len(flat_data["images"]) == file_data_size
+            assert len(flat_data["background"]) == file_data_size
+            for i in range(file_data_size):
+                file_datum: FileDatum = {}  # type: ignore
+                for key, _ in flat_data.items():
+                    file_datum[key] = (  # type: ignore
+                        {k: flat_data[key][k][i] for k in flat_data[key].keys()}
+                        if key == "tile_catalog"
+                        else flat_data[key][i]
+                    )
```

Comment:
the previous line is a bit hard to make sense of. Can it be simplified?

Reply:
Done.